File: leetcode/c++/linked_list_cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == nullptr) return false;
        if(head->next == nullptr) return false;
        
        ListNode *p = head;
        ListNode *q = head->next->next;
        
        while(q && p != q){
            p = p->next;
            if(q->next) q = q->next->next;
            else break;
        }
        
        if(p == q) return true;
        else return false;
    }
};

File: leetcode/java/LinkedListCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null || head.next == null) return false;
        
        ListNode p = head;
        ListNode q = head.next.next;
        
        while(q != null) {
            if(p == q) return true;
            p = p.next;
            if(q.next == null) return false;
            q = q.next.next;
        }
        return false;
    }
}
